### PR TITLE
Additional email validation forms

### PR DESCRIPTION
**Pull request by adriana-alter**
Goes with https://github.com/NYPL-Simplified/library_registry/tree/validate-emails-to-register (which will have its own PR as soon as I'm done with the tests).  Resolves https://jira.nypl.org/browse/SIMPLY-2605

Screen Shot 2020-03-05 at 4 40 40 PM
Screen Shot 2020-03-05 at 4 40 49 PM
Screen Shot 2020-03-05 at 4 41 33 PM
Screen Shot 2020-03-09 at 4 59 42 PM

### CODE REVIEW

**Comment by EdwinGuzman** [comment]
There are three forms and if they all have the same email, when you validate one form, the email gets validated and only that form tells you that it was updated. When you refresh the page, however, the timestamps for all the three forms are updated. This makes sense since it's the same email but it's a little bit odd since the other forms were explicitly validated.

I haven't been able to test this scenario but I don't expect it to break (and can you confirm it): if each form has a different email and one is validated, the timestamp for the other two forms shouldn't be updated.

File: src/components/EmailValidationForm.tsx [hunk 1]
```diff
@@ -1,71 +1,88 @@
 import * as React from "react";
-import { Form } from "library-simplified-reusable-components";
+import { Form, Button } from "library-simplified-reusable-components";
```
Comment (EdwinGuzman) [inline src/components/EmailValidationForm.tsx]: You're not using `Button`. Also, you can remove line 8, `LibraryData` isn't used.

File: src/components/EmailValidationForm.tsx [hunk 2]
```diff
@@ -1,71 +1,88 @@
...
 
-    let buttonText = hasEmail ? "Validate email address" : "No email address configured";
-    let buttonContent = <span>{buttonText}{icon}</span>;
-    let infoText = (alreadyValidated && !this.state.sent && !this.props.error) ? "Already validated" : null;
-    let successText = `Successfully validated ${this.props.library.urls_and_contact.contact_email}`;
+  renderInfoText(emailString: string, emailAddress: string, validated: string, alreadyValidated: boolean): string {
+    if (!emailAddress) {
+      return `No ${emailString} configured`;
+    } else if (!alreadyValidated) {
+      return "Not validated";
```
Comment (EdwinGuzman) [inline src/components/EmailValidationForm.tsx]: Do you have a screenshot of this to add to the PR just as an example?

File: src/components/EmailValidationForm.tsx [hunk 3]
```diff
@@ -1,71 +1,88 @@
...
+    } else if (!this.state.validated && !this.props.error) {
+      return `Validated: ${validated}`;
+    }
+  }
 
+  render(): JSX.Element {
+    let emailString = this.convert(this.props.email);
+    let emailAddress = this.props.libraryInfo[this.props.email];
+    let validated = this.props.libraryInfo[`${this.props.email.split("_")[0]}_validated`];
```
Comment (EdwinGuzman) [inline src/components/EmailValidationForm.tsx]: I'm guessing that the `libraryInfo` object has a property like `contact_validated` which is what you are getting here? Some documentation on that would be helpful since it's not easily readable or easy to tell why you're doing this.

File: src/components/EmailValidationForm.tsx [hunk 4]
```diff
@@ -1,71 +1,88 @@
...
-    let buttonText = hasEmail ? "Validate email address" : "No email address configured";
-    let buttonContent = <span>{buttonText}{icon}</span>;
-    let infoText = (alreadyValidated && !this.state.sent && !this.props.error) ? "Already validated" : null;
-    let successText = `Successfully validated ${this.props.library.urls_and_contact.contact_email}`;
+  renderInfoText(emailString: string, emailAddress: string, validated: string, alreadyValidated: boolean): string {
+    if (!emailAddress) {
+      return `No ${emailString} configured`;
+    } else if (!alreadyValidated) {
+      return "Not validated";
+    } else if (!this.state.validated && !this.props.error) {
```
Comment (EdwinGuzman) [inline src/components/EmailValidationForm.tsx]: Is this right? If it's not validated and there's no error, then display "validated"?

Also, what if there is an email, it's not already validated, and this last `else if` fails? There's no default return case.

Comment (EdwinGuzman) [reply src/components/EmailValidationForm.tsx]: Hmmm this makes sense because if `Form`'s `infoText` is undefined, nothing gets rendered, but if you didn't already know that or if `Form` gets updated in the future, you may want to be explicit about what is returned so there's no error.

File: src/components/EmailValidationSection.tsx [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import EmailValidationForm from "./EmailValidationForm";
+
+export interface EmailValidationSectionProps {
+  library: LibraryData;
+  store: Store<State>;
+}
+
+export default class EmailValidationSection extends React.Component<EmailValidationSectionProps, {}> {
+  render(): JSX.Element {
+    let emailForms = ["contact", "help", "copyright"].map((x) =>
```
Comment (EdwinGuzman) [inline src/components/EmailValidationSection.tsx]: A comment around the context for why these three email values are being validated would be nice for developers.

File: src/components/EmailValidationForm.tsx [hunk 6]
```diff
@@ -1,71 +1,88 @@
...
+    this.validate = this.validate.bind(this);
+    this.renderInfoText = this.renderInfoText.bind(this);
+    this.state = { validated: false };
   }
 
-  async sendEmail(data: FormData): Promise<void> {
-    await this.props.validateEmail(data);
-    await this.props.fetchLibrary(this.props.library.uuid);
-    this.setState({ sent: !this.props.error });
+  convert(email: string): string {
```
Comment (EdwinGuzman) [inline src/components/EmailValidationForm.tsx]: Why is this needed? It seems like it's manipulating emails from the server?

File: src/components/EmailValidationSection.tsx [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+      <li className="well">
+        <EmailValidationForm
+          store={this.props.store}
+          uuid={this.props.library.uuid}
+          email={`${x}_email`}
+          libraryInfo={this.props.library.urls_and_contact}
```
Comment (EdwinGuzman) [inline src/components/EmailValidationSection.tsx]: I'm not sure if rendering one form three times with the same "hiddenName" and "hiddenValue" is a problem. Probably not but something I can look into later.

File: src/components/EmailValidationForm.tsx (hunk 6 above)
Comment (adriana-alter) [reply src/components/EmailValidationForm.tsx]: No, it's converting the name of the email, not the actual email address.  Like, turning "contact_email" into "contact email" and vice versa.

Comment (EdwinGuzman) [reply src/components/EmailValidationForm.tsx]: I see. Maybe `convertEmailTitle` would be better or something like that?

File: src/components/EmailValidationForm.tsx (hunk 4 above)
Comment (adriana-alter) [reply src/components/EmailValidationForm.tsx]: Yes, that's right.  `this.state.validated` isn't whether the email has previously been validated, it's whether you've clicked "validate" just now.  So:
-- If it's never been validated, it displays "not validated".
-- If it has been validated at some point in the past, but not by you having clicked it just now, it displays "validated".
-- If it's validated because you just clicked it, i.e. if `this.state.validated` is true, it displays "successfully validated" and turns green (that's the `successText` being passed to the Form).

File: src/components/EmailValidationSection.tsx (hunk 7 above)
Comment (adriana-alter) [reply src/components/EmailValidationSection.tsx]: I'm not finding anything online that suggests it would be a problem.  I can see how there would be issues if all the forms were somehow being submitted together, but since they only get submitted one at a time...

File: src/components/EmailValidationForm.tsx (hunk 4 above)
Comment (EdwinGuzman) [reply src/components/EmailValidationForm.tsx]: So, if you just clicked to validate it, `this.state.validated` becomes true and the `successText` is shown, that I get. But this last `else if` is checking that `this.state.validated` is not true and there's no error. Is it assuming that it was therefore already validated at `validated` time stamp?

**Comment by adriana-alter** [comment]
I haven't been able to test the they-each-have-different-emails scenario either; I'll deal with it on Monday!  (I don't see any way it would break, but...)

File: src/components/EmailValidationForm.tsx (hunk 4 above)
Comment (adriana-alter) [reply src/components/EmailValidationForm.tsx]: Yes--that's the only other possibility.  We've already covered "there's no email address" and "there's an email address but it's never been validated."

File: src/components/EmailValidationSection.tsx (hunk 7 above)
Comment (EdwinGuzman) [reply src/components/EmailValidationSection.tsx]: Yeah, since each is its own form, there shouldn't be a problem for individual submissions. But I'm not sure how screenreaders handle that (if they read the whole page and see the same value three times or read individual forms and see the value once). In any case, nothing to worry about at the moment.

**Comment by adriana-alter** [comment]
Totally fine if they have different emails!

**Comment by EdwinGuzman** [comment]
Cool. So the previous comment

> There are three forms and if they all have the same email, when you validate one form, the email gets validated and only that form tells you that it was updated. When you refresh the page, however, the timestamps for all the three forms are updated. This makes sense since it's the same email but it's a little bit odd since the other forms were explicitly validated.

is this how it's suppose to work?

What did the last commit fix? Also, just minor comments about writing comments and you're all set.

**Comment by adriana-alter** [comment]
That is how it's supposed to work, if they all have the same email address.  I agree it's a little bit odd, but all the alternatives I can think of seem worse.  Do you have an idea?

The last commit is to fix my discovery that when they have different email addresses and you validate one of them, the outdated info text AND the success text get displayed, like so:
Screen Shot 2020-03-09 at 1 22 48 PM

**Comment by adriana-alter** [comment]
I'm pretty sure I added comments to all the places that you pointed out should have one; did I miss something?  (Or did you just mean that you're going to write more comments now?  The caffeine never quite kicked in...)

File: src/components/EmailValidationForm.tsx (hunk 2 above)
Comment (EdwinGuzman) [reply src/components/EmailValidationForm.tsx]: @adriana-alter here's an unanswered question.

File: src/components/EmailValidationForm.tsx (hunk 3 above)
Comment (EdwinGuzman) [reply src/components/EmailValidationForm.tsx]: @adriana-alter here too.

File: src/components/EmailValidationSection.tsx (hunk 5 above)
Comment (EdwinGuzman) [reply src/components/EmailValidationSection.tsx]: @adriana-alter

**Comment by EdwinGuzman** [comment]
☕️ ☕️ ☕️ 

I'm not sure how to approach that... but it's small enough that maybe it's not a big issue? When you refresh all the old values get updated and can cause confusion, but maybe it's really not a problem. I'll approve but did note where you could use some more comments.

File: src/components/EmailValidationForm.tsx (hunk 2 above)
Comment (adriana-alter) [reply src/components/EmailValidationForm.tsx]: Ah.  The answer was "no."  I'll generate one--hang on...

File: src/components/EmailValidationForm.tsx (hunk 3 above)
Comment (adriana-alter) [reply src/components/EmailValidationForm.tsx]: I did; github is just being weird.
Screen Shot 2020-03-09 at 5 01 56 PM

File: src/components/EmailValidationSection.tsx (hunk 5 above)
Comment (adriana-alter) [reply src/components/EmailValidationSection.tsx]: And again.  Github being weird.

Screen Shot 2020-03-09 at 5 05 52 PM
